python-oauth-web-app/app.py

```diff
@@ -0,0 +1,300 @@
+    if request.method != "POST":
+        return render_template("login.html")
+
+    # resolve handle and DID document, using atproto identity helpers
+    username = request.form["username"]
+    if not valid_handle(username):
+        flash("Invalid Handle")
+        return render_template("login.html"), 400
+    did = resolve_handle(username)
+    if not did:
+        flash("Handle Not Found")
+        return render_template("login.html"), 400
+    if not valid_did(did):
+        flash("Handle resolved to invalid DID")
+        return render_template("login.html"), 400
+    did_doc = resolve_did(did)
+    if not did_doc:
+        flash("DID Not Found: " + did)
+        return render_template("login.html"), 400
+
+    # fetch PDS OAuth metadata
+    # IMPORTANT: PDS endpoint URL is untrusted input, SSRF mitigations are needed
+    pds_url = pds_endpoint(did_doc)
+    print(f"PDS: {pds_url}")
+    assert is_safe_url(pds_url)
+    with hardened_http.get_session() as sess:
+        resp = sess.get(f"{pds_url}/.well-known/oauth-protected-resource")
+    resp.raise_for_status()
+
+    # IMPORTANT: Authorization Server URL is untrusted input, SSRF mitigations are needed
+    authsrv_url = resp.json()["authorization_servers"][0]
+    print(f"Auth Server: {authsrv_url}")
+    assert is_safe_url(authsrv_url)
+    try:
+        authsrv_meta = fetch_authsrv_meta(authsrv_url)
+    except Exception as err:
+        print(f"failed to fetch authsrv_meta: " + str(err))
+        flash("Failed to fetch Auth Server (Entryway) OAuth metadata")
+        return render_template("login.html"), 400
+
+    app_url = request.url_root.replace("http://", "https://")
+    client_id = f"{app_url}oauth/client-metadata.json"
+
+    pkce_verifier, state, nonce, resp = send_par_auth_request(
+        authsrv_url, authsrv_meta, did, app_url, secret_jwk
+    )
+    resp.raise_for_status()
+    request_uri = resp.json()["request_uri"]
+
+    # create a DPoP keypair for this user session now (TODO: why now?)
+    dpop_private_jwk = JsonWebKey.generate_key("EC", "P-256", is_private=True).as_json(
+        is_private=True
+    )
+
+    # persist auth request to database
+    print(f"saving oauth_auth_request to DB: {state}")
+    query_db(
+        "INSERT INTO oauth_auth_request (state, iss, nonce, pkce_verifier, dpop_private_jwk, did, handle, pds_url) VALUES(?, ?, ?, ?, ?, ?, ?, ?);",
+        [
+            state,
+            authsrv_meta["issuer"],
+            nonce,
+            pkce_verifier,
+            dpop_private_jwk,
+            did,
+            username,
+            pds_url,
+        ],
+    )
+
+    # IMPORTANT: Authorization endpoint URL is untrusted input, security mitigations are needed before redirecting user
+    auth_url = authsrv_meta["authorization_endpoint"]
+    assert is_safe_url(auth_url)
+    qparam = urlencode({"client_id": client_id, "request_uri": request_uri})
+    return redirect(f"{auth_url}?{qparam}")
+
+
+@app.route("/oauth/authorize")
```

Comment:
nit: `/authorize` is typically used by Authorization Server as the "authorization endpoint". the `redirect_uri` is more commonly called `/callback` or `/redirect`.

Reply:
updated to `/callback`